slackevents/parsers.go

```diff
@@ -0,0 +1,194 @@
+package slackevents
+
+import (
+	"crypto/subtle"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/ACollectionOfAtoms/slack"
+)
+
+// eventsMap checks both slack.EventsMapping and
+// and slackevents.EventsAPIInnerEventMapping. If the event
+// exists, returns the the unmarshalled struct instance of
+// target for the matching event type.
+// TODO: Consider moving all events into its own package?
+func eventsMap(t string) (interface{}, bool) {
+	// Must parse EventsAPI FIRST as both RTM and EventsAPI
+	// have a type: "Message" event.
+	// TODO: Handle these cases more explicitly.
+	v, exists := EventsAPIInnerEventMapping[t]
+	if exists {
+		return v, exists
+	}
+	v, exists = slack.EventMapping[t]
+	if exists {
+		return v, exists
+	}
+	return v, exists
+}
+
+func parseOuterEvent(rawE json.RawMessage) (EventsAPIEvent, error) {
+	e := &EventsAPIEvent{}
+	err := json.Unmarshal(rawE, e)
+	if err != nil {
+		return EventsAPIEvent{
+			"",
+			"unmarshalling_error",
+			&slack.UnmarshallingErrorEvent{ErrorObj: err},
+			EventsAPIInnerEvent{},
+		}, err
+	}
+	if e.Type == CallbackEvent {
+		cbEvent := &EventsAPICallbackEvent{}
+		err = json.Unmarshal(rawE, cbEvent)
+		if err != nil {
+			return EventsAPIEvent{
+				"",
+				"unmarshalling_error",
+				&slack.UnmarshallingErrorEvent{ErrorObj: err},
+				EventsAPIInnerEvent{},
+			}, err
+		}
+		return EventsAPIEvent{
+			e.Token,
+			e.Type,
+			cbEvent,
+			EventsAPIInnerEvent{},
+		}, nil
+	}
+	urlVE := &EventsAPIURLVerificationEvent{}
+	err = json.Unmarshal(rawE, urlVE)
+	if err != nil {
+		return EventsAPIEvent{
+			"",
+			"unmarshalling_error",
+			&slack.UnmarshallingErrorEvent{ErrorObj: err},
+			EventsAPIInnerEvent{},
+		}, err
+	}
+	return EventsAPIEvent{
+		e.Token,
+		e.Type,
+		urlVE,
+		EventsAPIInnerEvent{},
+	}, nil
+}
+
+func parseInnerEvent(e *EventsAPICallbackEvent) (EventsAPIEvent, error) {
+	iE := &slack.Event{}
+	rawInnerJSON := e.InnerEvent
+	err := json.Unmarshal(*rawInnerJSON, iE)
+	if err != nil {
+		return EventsAPIEvent{
+			e.Token,
+			"unmarshalling_error",
+			&slack.UnmarshallingErrorEvent{ErrorObj: err},
+			EventsAPIInnerEvent{},
+		}, err
+	}
+	v, exists := eventsMap(iE.Type)
+	if !exists {
+		return EventsAPIEvent{
+			e.Token,
+			iE.Type,
+			nil,
+			EventsAPIInnerEvent{},
+		}, fmt.Errorf("Inner Event does not exist! %s", iE.Type)
+	}
+	t := reflect.TypeOf(v)
+	recvEvent := reflect.New(t).Interface()
+	err = json.Unmarshal(*rawInnerJSON, recvEvent)
+	if err != nil {
+		return EventsAPIEvent{
+			e.Token,
+			"unmarshalling_error",
+			&slack.UnmarshallingErrorEvent{ErrorObj: err},
+			EventsAPIInnerEvent{},
+		}, err
+	}
+	return EventsAPIEvent{
+		e.Token,
+		e.Type,
+		e,
+		EventsAPIInnerEvent{iE.Type, recvEvent},
+	}, nil
+}
+
+type Config struct {
+	VerificationToken string
```

Comment:
I think the config should just contain the verifier type atm. but can be discussed later.